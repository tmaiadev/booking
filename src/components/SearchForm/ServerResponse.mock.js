export default {
	results: {
		isGooglePowered: !1,
		numFound: 578,
		docs: [{
			placeType: "A",
			placeKey: "1472178",
			locationId: "37766",
			name: "Heathrow Airport",
			iata: "LHR",
			city: "London",
			region: "Greater London",
			country: "United Kingdom",
			countryIso: "gb",
			ufi: 900038535,
			bookingId: "airport-37766",
			lat: 51.47079849243164,
			lng: -.45304301381111145,
			alternative: ["GB,UK,England,Heathrow"],
			searchType: "L",
			lang: "en",
			index: 1,
			isPopular: !0
		}, {
			placeType: "C",
			placeKey: "441379",
			locationId: "20866",
			name: "London",
			region: "Greater London",
			country: "United Kingdom",
			countryIso: "gb",
			ufi: -2601889,
			bookingId: "city-2623186",
			lat: 51.500999450683594,
			lng: -.1263429969549179,
			alternative: ["GB,UK,England"],
			searchType: "L",
			lang: "en",
			index: 2,
			isPopular: !1
		}, {
			placeType: "A",
			placeKey: "1472185",
			locationId: "38106",
			name: "London Luton Airport",
			iata: "LTN",
			city: "London",
			region: "Greater London",
			country: "United Kingdom",
			countryIso: "gb",
			ufi: 900038596,
			bookingId: "airport-38106",
			lat: 51.87990188598633,
			lng: -.3771649897098541,
			alternative: ["GB,UK,England,Luton"],
			searchType: "L",
			lang: "en",
			index: 3,
			isPopular: !0
		}, {
			placeType: "A",
			placeKey: "1472186",
			locationId: "38111",
			name: "London Stansted Airport",
			iata: "STN",
			city: "London",
			region: "Greater London",
			country: "United Kingdom",
			countryIso: "gb",
			ufi: 900038566,
			bookingId: "airport-38111",
			lat: 51.889400482177734,
			lng: .2615149915218353,
			alternative: ["GB,UK,England,Stansted"],
			searchType: "L",
			lang: "en",
			index: 4,
			isPopular: !0
		}, {
			placeType: "A",
			placeKey: "1472173",
			locationId: "37761",
			name: "Gatwick Airport",
			iata: "LGW",
			city: "London",
			region: "Greater London",
			country: "United Kingdom",
			countryIso: "gb",
			ufi: 900038541,
			bookingId: "airport-37761",
			lat: 51.15449905395508,
			lng: -.16539999842643738,
			alternative: ["GB,UK,England,Gatwick"],
			searchType: "L",
			lang: "en",
			index: 5,
			isPopular: !0
		}, {
			placeType: "T",
			placeKey: "992437016",
			locationId: "38056",
			name: "London - Victoria Station",
			city: "London",
			region: "England",
			country: "United Kingdom",
			countryIso: "gb",
			ufi: 23862,
			bookingId: "train-23862",
			lat: 51.49509811401367,
			lng: -.1441899985074997,
			alternative: ["GB,UK"],
			searchType: "L",
			lang: "en",
			index: 6,
			isPopular: !1
		}, {
			placeType: "A",
			placeKey: "1472184",
			locationId: "38101",
			name: "London City Airport",
			iata: "LCY",
			city: "London",
			region: "Greater London",
			country: "United Kingdom",
			countryIso: "gb",
			ufi: 900038685,
			bookingId: "airport-38101",
			lat: 51.503700256347656,
			lng: .05320800095796585,
			alternative: ["GB,UK,England,London City"],
			searchType: "L",
			lang: "en",
			index: 7,
			isPopular: !1
		}]
	}
};