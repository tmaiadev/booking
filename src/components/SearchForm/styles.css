.search-form {
    --border-radius: 6px;

    background-color: var(--surface-color-2);
    border-radius: var(--border-radius);
    display: flex;
    flex-direction: column;
    gap: 4px;
    padding: 4px;
}

.search-form .input__input {
    border-radius: var(--border-radius) var(--border-radius) 0 0;
}

.search-form .search-form__submit-button {
    border-radius: 0 0 var(--border-radius) var(--border-radius);
}

.search-form__submit-button {
    background-color: var(--surface-color-3);
    border-width: 0;
    box-sizing: border-box;
    color: var(--text-color);
    cursor: pointer;
    font-size: 1.4em;
    font-weight: 700;
    height: 64px;
    padding: 8px 16px;
    transition: background-color ease-in-out .1s, color ease-in-out .1s;
}

.search-form__submit-button:hover,
.search-form__submit-button:focus {
    background-color: var(--surface-color-3-light);
}

.search-form__submit-button:active {
    background-color: var(--surface-color-3-dark);
    color: var(--text-color-2);
}

@media (min-width: 1024px) {
    .search-form {
        flex-direction: row;
    }

    .search-form > :first-child {
        flex-basis: 100%;
    }

    .search-form .input__input {
        border-radius: var(--border-radius) 0 0 var(--border-radius);
    }
    
    .search-form .search-form__submit-button {
        border-radius: 0 var(--border-radius) var(--border-radius) 0;
    }

    [dir="rtl"] .search-form .input__input {
        border-radius: 0 var(--border-radius) var(--border-radius) 0;
        flex-basis: 100%;
    }
    
    [dir="rtl"] .search-form .search-form__submit-button {
        border-radius: var(--border-radius) 0 0 var(--border-radius);
    }
}

@media (prefers-reduced-motion) {
    .search-form__submit-button {
        transition: none;
    }
}

@media (forced-colors: active) {
    .search-form__submit-button {
        border-width: initial;
    }
}