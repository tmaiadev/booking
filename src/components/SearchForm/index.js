import { useState, useEffect } from 'react';
import { endpoint, mapEndpointResultsToDataList } from '../../helpers/locationLookup';
import Input from '../Input';
import './styles.css';

const Form = () => {
    const [pickupLocationQuery, setPickupLocationQuery] = useState('');
    const [dataList, setDataList] = useState([]);
    const [loading, setLoading] = useState(false);

    const onSubmit = evt => evt.preventDefault();

    useEffect(() => {
        if (pickupLocationQuery.length <= 1) {
            setDataList([]);

            return;
        }

        setLoading(true);

        fetch(endpoint(pickupLocationQuery))
            .then(data => data.json())
            .then(data => setDataList(data?.results?.docs.map(mapEndpointResultsToDataList) || []))
            .catch(e => setDataList([]))
            .finally(() => setLoading(false))

    }, [pickupLocationQuery]);

    return <form className="search-form" onSubmit={onSubmit}>
        <Input
            id="pickup-location"
            placeholder="Pick-up Location"
            onChange={setPickupLocationQuery}
            debounce={500}
            loading={loading}
            dataList={dataList}
        />
        <button type="submit" className="search-form__submit-button">
            Search
        </button>
    </form>
};

export default Form;