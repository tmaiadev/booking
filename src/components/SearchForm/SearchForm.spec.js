import { render, fireEvent, waitFor } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import SearchForm from './';
import serverResponseMock from './ServerResponse.mock';

jest.spyOn(global, 'fetch');
jest.useFakeTimers();

describe('Given the `SearchForm` component', () => {
    it('should render successfully', () => {
        const { container } = render(<SearchForm />);

        expect(container).toMatchSnapshot();
    });

    describe('When the user starts typing on the Pick-up Location input', () => {
        describe('and user types less than 2 characters', () => {
            it('should NOT make a network request', () => {
                fetch.mockResolvedValue(true);

                const { getAllByLabelText } = render(<SearchForm />);
                const [input] = getAllByLabelText('Pick-up Location');

                fireEvent.change(input, { target: { value: 'L' }});

                act(() => {
                    jest.advanceTimersByTime(600);
                });

                expect(fetch).not.toHaveBeenCalled();
            });
        });

        describe('and user types more than 2 characters', () => {
            let renderer, input;

            beforeEach(async () => {
                fetch.mockResolvedValue({
                    json: () => Promise.resolve(serverResponseMock)
                });

                renderer = render(<SearchForm />);
                const { getAllByLabelText, queryByTestId } = renderer;
                input = getAllByLabelText('Pick-up Location')[0];

                fireEvent.change(input, { target: { value: 'London Airport' }});

                act(() => {
                    jest.advanceTimersByTime(600);
                });

                await waitFor(() => expect(queryByTestId('loader')).toBeInTheDocument());
                await waitFor(() => expect(queryByTestId('loader')).not.toBeInTheDocument());
            });

            it('should request the correct url', () => {
                expect(fetch).toHaveBeenCalledWith('https://www.rentalcars.com/FTSAutocomplete.do?solrIndex=fts_en&solrRows=6&solrTerm=London%20Airport');
            });

            it('should render a combobox', async () => {
                serverResponseMock.results.docs.forEach(({ name }) => 
                    expect(renderer.queryByText(name)).toBeInTheDocument());
            });

            describe('and user selects a combobox suggestion', () => {
                it('should update the input value', () => {
                    const option = renderer.queryByText('Heathrow Airport');

                    fireEvent.click(option);

                    expect(input.value).toBe('Heathrow Airport, London, Greater London, United Kingdom');
                });
            });
        });
    });
});