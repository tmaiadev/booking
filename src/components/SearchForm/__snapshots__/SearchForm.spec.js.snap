// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the \`SearchForm\` component should render successfully 1`] = `
<div>
  <form
    class="search-form"
  >
    <div
      class="input"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-owns="pickup-location-listbox"
        role="combobox"
      >
        <input
          aria-autocomplete="list"
          aria-controls="pickup-location-listbox"
          aria-label="Pick-up Location"
          class="input__input"
          id="pickup-location"
          value=""
        />
      </div>
      <ul
        aria-labelledby="pickup-location-label"
        class="input__listbox input__listbox--hidden"
        id="pickup-location-listbox"
        role="listbox"
      />
      <div
        aria-hidden="true"
        class="input__container"
      >
        <svg
          class="input__icon"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
        <label
          class="input__label"
          for="pickup-location"
          id="pickup-location-label"
        >
          Pick-up Location
        </label>
      </div>
    </div>
    <button
      class="search-form__submit-button"
      type="submit"
    >
      Search
    </button>
  </form>
</div>
`;
